### ISSUE TITLE
refresh: Error applying view update to 10.142.0.5: exceptions::mutation_write_timeout_exception

### ISSUE THREAD

**Issue by amoskong** [scylladb/scylla#3080]
*Installation details*
Scylla version (or git commit hash): 2.0.2-0.20171201.07b039f
Cluster size: 6
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS

Jenkins job: [scylla-2.1-longevity-gce](http://jenkins.cloudius-systems.com:8080/job/scylla-master-longevity-gce/label=gce3-qavpc/36/)

In longevity we have 5 worklodes run in parallel:
- "cassandra-stress mixed cl=QUORUM duration=60m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=20 -pop seq=1..100000000 -log interval=5"
- "cassandra-stress counter_write cl=QUORUM duration=60m -schema 'replication(factor=3) compaction(strategy=DateTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=1 -pop seq=1..1000000"
- "cassandra-stress user profile=/tmp/cs_mv_profile.yaml ops'(insert=3,read1=1,read2=1,read3=1)' cl=QUORUM duration=60m -port jmx=6868 -mode cql3 native -rate threads=10"
- "cassandra-stress read duration=60m -port jmx=6868 -mode cql3 native -rate threads=20 -pop seq=1..100000000 -log interval=5"
- "cassandra-stress counter_read duration=60m -port jmx=6868 -mode cql3 native -rate threads=10 -pop seq=1..1000000"

During the refresh nemesis, both database and stress client raised some error, and `counter_write,  user profile and counter_read` workload FAILURE in the end (reached to 60 mins).
`mixed and read` workload are successfully END.

Majory logs from SCT:

```
06:27:19,363 | sdcm.nemesis.ChaosMonkey: <<<<<<<<<<<<<Finished random_disrupt_method nodetool_enospc

06:34:08,458 | sdcm.nemesis.ChaosMonkey: >>>>>>>>>>>>>Started random_disrupt_method nodetool_refresh

06:34:14,733 | Remote [scylla-test@10.142.0.5]: Running 'nodetool --host localhost refresh -- keyspace1 standard1'

06:34:18,136 | [10.142.0.5] [stdout] Dec 20 06:34:18 info   |  [shard 12] storage_service - Loading new SSTables for keyspace1.standard1...
06:34:18,151 | [10.142.0.5] [stdout] Dec 20 06:34:18 info   |  [shard 12] storage_service - CF keyspace1.standard1 at shard 12 had SSTables writes disabled for 14582 usec

06:34:18,161 | [10.142.0.5] [stdout] Dec 20 06:34:18 info   |  [shard 15] storage_service - CF keyspace1.standard1 at shard 15 had SSTables writes disabled for 15239 usec

06:34:20,218 | [10.142.0.3] [stdout] Dec 20 06:34:20 err    |  [shard 11] view - Error applying view update to 10.142.0.5: exceptions::mutation_write_timeout_exception (Operation timed out for mview.users_by_first_name - received only 0 responses from 1 CL=ONE.)

06:34:20,254 | [10.142.0.4] [stdout] Dec 20 06:34:20 err    |  [shard 4] view - Error applying view update to 10.142.0.5: exceptions::mutation_write_timeout_exception (Operation timed out for mview.users_by_last_name - received only 0 responses from 1 CL=ONE.)

06:34:22,272 | [10.142.0.5] [stdout] Dec 20 06:34:22 info   | Reactor stalled for 4000 ms on shard 13.
06:34:22,272 | [10.142.0.5] [stdout] Dec 20 06:34:22 info   | Backtrace:

06:34:23,255 | [10.142.0.27] [stdout] com.datastax.driver.core.exceptions.ReadTimeoutException: Cassandra timeout during read query at consistency QUORUM (3 responses were required but only 2 replica responded)
06:34:23,257 | [10.142.0.27] [stdout] java.util.NoSuchElementException
06:34:23,269 | [10.142.0.27] [stdout] java.io.IOException: Operation x10 on key(s) [\us H-^^&^ZxBT]: Error executing: (NoSuchElementException)

06:34:26,112 | sdcm.nemesis.ChaosMonkey: <<<<<<<<<<<<<Finished random_disrupt_method nodetool_refresh
```

**Comment by amoskong** [comment]
About refresh nemesis:
- download https://s3.amazonaws.com/scylla-qa-team/keyspace1.standard1.100M.tar.gz
- uncompress the file to `/var/lib/scylla/data/keyspace1/xxx/upload/`
- nodetool --host localhost refresh -- keyspace1 standard1
- verify the data is loaded: select * from keyspace1.standard1 where key=0x314e344b4d504d4b4b30

**Comment by amoskong** [comment]
Snapshot during the refresh nemesis:
[fireshot capture 2 - grafana - scylla per server metrics 2 _ - http___35 196 95 169_3000_dashboard](https://user-images.githubusercontent.com/309708/34202494-df616bb8-e5b2-11e7-947c-8f24c40922f1.png)

Snapshot of whole longevity:
[fireshot capture 1 - grafana - scylla per server metrics 2 _ - http___35 196 95 169_3000_dashboard](https://user-images.githubusercontent.com/309708/34202311-6399d0a6-e5b2-11e7-8abd-df7fd25ea7ff.png)

**Comment by duarten** [comment]
This seems expected if the target node was down or overloaded. What exactly is to be fixed here? /cc @nyh

**Comment by juliayakovlev** [comment]
The problem happens when  there are 4 and more nodes in the cluster and one node (or more) is stopped/removed during inserts into MV.
The tests are added to the dtest.
Pull request waiting for review.

**Comment by roydahan** [comment]
This is probably the same issue as #2783.

Anyway, @juliayakovlev added several dtest cases that reproduce it easily: stop_node_during_mv_insert_4_nodes_test
remove_node_during_mv_insert_4_nodes_test
double_node_failure_during_mv_insert_4_nodes_test

**Comment by duarten** [comment]
Maybe @nyh can look into this, I think he wrote the relevant code here.

**Comment by slivne** [comment]
reassigned @nyh

**Comment by duarten** [comment]
This seems indeed like it is a duplicate of #2783, but I don't understand what the issue is. Looking at the dtests, they remove a node in the middle of a c-s run, so of course writes are going to timeout - a base replica is trying to talk to a view replica, but the view replica is gone without warning.

**Comment by duarten** [comment]
The reason this happens for a 4 node cluster is because the nodes are not simply paired with themselves, as is the case in a 3 node cluster with RF=3.

Regarding the dtests, I think they can be removed - they are not even asserting that the MVs contain a set of rows, they just assert on the log file containing no errors.

**Comment by juliayakovlev** [comment]
@duarten  I do not see the reason to remove tests. I'll add additional assertions before checking the log. But we still need them

**Comment by roydahan** [comment]
@duarten, the issue is / was (does it still reproduce?) that the c-s load for that user profile stops right after the timeouts.

**Comment by duarten** [comment]
@roydahan That's an issue with cassandra-stress, no? It expects all values to have been inserted when it does the reading round.

@juliayakovlev I don't think they have any value, and they will consistently fail.

**Comment by duarten** [comment]
@roydahan Yes, that's what I wrote the in my comment above. The write timeout is expected if the target view replica is taken down (stop node, remove node) without any changed to the ring. 

Aren't the `java.util.NoSuchElementException` exceptions that pop up in c-s related to it expecting that all writes succeeded?

**Comment by roydahan** [comment]
I see now also the read exception, not sure why it happens:
`ReadTimeoutException: Cassandra timeout during read query at consistency QUORUM (3 responses were required but only 2 replica responded)
`

The cluster is 4 nodes, rf=3, cl=Quorum.
Why is it expecting 3 responses for Quorum?

**Comment by duarten** [comment]
I don't know! I asked the same question over at #2704, which I think is the same issue. Waiting for @gleb-cloudius to shed some light.

**Comment by duarten** [comment]
@roydahan That exception is coming from the Java driver. So either they have a bug, or the RF is 4 instead of 3.

**Comment by duarten** [comment]
Oh, it can be for reconciliation. For some reason, when we reconcile, we asked for data from all targets. Is this really so, @gleb-cloudius? If it is, there might be more replicas in the targets (extra replica, read repair replica) then the ones necessary to meet the CL, and more replicas then the ones that answered originally; seems wrong.

**Comment by duarten** [comment]
Spoke with @gleb-cloudius. The extra replica can be due to foreground read repair or foreground reconciliation.

To sum things up: the errors happen naturally (a node is being stopped, and other nodes try to write to it);  those tests should expect errors in the log file. However, if they read the data with consistency > ONE, then all results should be there.

**Comment by duarten** [comment]
Alright, closing this as not an issue. The relevant tests (i.e., `stop_node_during_mv_insert_4_nodes_test`) are broken, because errors will always appear in the logs if a node is going down and others are trying to send view updates to it. As for the `cassandra-stress` run, it will only be successful when we have consistency mechanisms (e.g., repair) to ensure the missed view updates will eventually propagate there, such that during the read phase, `cassandra-stress` will find everything there. /cc @slivne

**Comment by juliayakovlev** [comment]
It's relevant test. This situation may happen for the customers. I'll add additional validations.
@slivne Shlomi, can I ignore this error? If yes, should it be error? Maybe it's warning?
@roydahan Roy, even I'll ignore this error, dtest will fail the test after it'll be completed because of this error will found in the log. Is there some way to say to the dtest ignore it after finish of the test?

**Comment by roydahan** [comment]
I’m still not convinced this is not an issue.
We need to discuss it.

**Comment by duarten** [comment]
@juliayakovlev I didn't say there weren't relevant, I said they're broken. You could adapt the `mv_populating_from_existing_*` tests, which cover these scenarios, for the case where the view is already populated. The error message should be an error message, it is just expected to happen in this particular case. Those tests will fail not only because of the error messages (`self.allow_log_errors = True`), but also because of how cassandra-stress behaves I think.

**Comment by duarten** [comment]
@roydahan Why? What am I missing?
